File: implementations/python-sdk/aether/filters.py
"""Filter matching for client-side subscriptions."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Iterable, Mapping

from .crypto import Tag, normalize_tags


@dataclass(frozen=True)
class EventFilter:
    kinds: set[int] | None = None
    pubkey_prefixes: tuple[bytes, ...] | None = None
    tags: set[tuple[str, str]] | None = None
    since: int | None = None
    until: int | None = None


def normalize_filter(raw: Mapping[str, object]) -> EventFilter:
    kinds = _normalize_kinds(raw.get("kinds"))
    pubkey_prefixes = _normalize_pubkey_prefixes(raw.get("pubkey_prefixes"))
    tags = _normalize_tag_filters(raw.get("tags"))
    since = _normalize_int(raw.get("since"))
    until = _normalize_int(raw.get("until"))
    return EventFilter(
        kinds=kinds,
        pubkey_prefixes=pubkey_prefixes,
        tags=tags,
        since=since,
        until=until,
    )


def match_event(event: Mapping[str, object], flt: EventFilter) -> bool:
    kind = _parse_int(event.get("kind"), "kind")
    created_at = _parse_int(event.get("created_at"), "created_at")
    pubkey = _parse_hex_or_bytes(event.get("pubkey"), "pubkey", 32)
    tags = normalize_tags(_parse_tags(event.get("tags")))

    if flt.kinds is not None and kind not in flt.kinds:
        return False
    if flt.pubkey_prefixes is not None and not _match_pubkey_prefix(pubkey, flt.pubkey_prefixes):
        return False
    if flt.tags is not None and not _match_tags(tags, flt.tags):
        return False
    if flt.since is not None and created_at < flt.since:
        return False
    if flt.until is not None and created_at > flt.until:
        return False
    return True


def _normalize_kinds(raw: object) -> set[int] | None:
    if raw is None:
        return None
    if not isinstance(raw, (list, tuple, set)):
        raise ValueError("kinds must be a list")
    return {_parse_int(value, "kind") for value in raw}


def _normalize_pubkey_prefixes(raw: object) -> tuple[bytes, ...] | None:
    if raw is None:
        return None
    if not isinstance(raw, (list, tuple)):
        raise ValueError("pubkey_prefixes must be a list")
    prefixes = []
    for value in raw:
        prefix = _parse_hex_or_bytes(value, "pubkey_prefix", 16)
        prefixes.append(prefix)
    return tuple(prefixes)


def _normalize_tag_filters(raw: object) -> set[tuple[str, str]] | None:
    if raw is None:
        return None
    tags: set[tuple[str, str]] = set()
    if isinstance(raw, dict):
        for key, values in raw.items():
            if not isinstance(key, str):
                raise ValueError("tag key must be str")
            if not isinstance(values, (list, tuple, set)):
                raise ValueError("tag values must be list")
            for value in values:
                tags.add((key, str(value)))
        return tags
    if not isinstance(raw, (list, tuple, set)):
        raise ValueError("tags must be list or dict")
    for entry in raw:
        if isinstance(entry, (list, tuple)) and len(entry) == 2:
            key, value = entry
            if not isinstance(key, str):
                raise ValueError("tag key must be str")
            tags.add((key, str(value)))
        else:
            raise ValueError("tag filter entries must be (key, value)")
    return tags


def _match_pubkey_prefix(pubkey: bytes, prefixes: Iterable[bytes]) -> bool:
    return any(pubkey.startswith(prefix) for prefix in prefixes)


def _match_tags(tags: Iterable[Tag], required: set[tuple[str, str]]) -> bool:
    available = {(tag.key, value) for tag in tags for value in tag.values}
    return required.issubset(available)


def _parse_hex_or_bytes(value: object, field: str, size: int) -> bytes:
    if isinstance(value, bytes):
        data = value
    elif isinstance(value, str):
        try:
            data = bytes.fromhex(value)
        except ValueError as exc:
            raise ValueError(f"{field} must be hex") from exc
    else:
        raise ValueError(f"{field} must be bytes or hex string")

    if len(data) != size:
        raise ValueError(f"{field} must be {size} bytes")
    return data


def _parse_int(value: object, field: str) -> int:
    if isinstance(value, bool) or value is None:
        raise ValueError(f"{field} must be int")
    if isinstance(value, int):
        return value
    if isinstance(value, str):
        try:
            return int(value)
        except ValueError as exc:
            raise ValueError(f"{field} must be int") from exc
    raise ValueError(f"{field} must be int")


def _normalize_int(value: object) -> int | None:
    if value is None:
        return None
    return _parse_int(value, "filter")


def _parse_tags(value: object) -> list[object]:
    if value is None:
        return []
    if not isinstance(value, (list, tuple)):
        raise ValueError("tags must be a list")
    return list(value)

File: sdk/typescript/src/filters.ts
export type EventFilter = {
  kinds?: Set<number>;
  pubkeyPrefixes?: Uint8Array[];
  tags?: Set<string>;
  since?: number;
  until?: number;
};

export function normalizeFilter(raw: Record<string, unknown>): EventFilter {
  const kinds = normalizeKinds(raw.kinds);
  const pubkeyPrefixes = normalizePubkeyPrefixes(raw.pubkey_prefixes);
  const tags = normalizeTags(raw.tags);
  const since = normalizeInt(raw.since);
  const until = normalizeInt(raw.until);
  return { kinds, pubkeyPrefixes, tags, since, until };
}

export function matchEvent(event: Record<string, unknown>, filter: EventFilter): boolean {
  const kind = parseIntField(event.kind, "kind");
  const createdAt = parseIntField(event.created_at, "created_at");
  const pubkey = parseHexOrBytes(event.pubkey, "pubkey", 32);
  const tags = parseTags(event.tags);

  if (filter.kinds && !filter.kinds.has(kind)) {
    return false;
  }
  if (filter.pubkeyPrefixes && !filter.pubkeyPrefixes.some((prefix) => startsWith(pubkey, prefix))) {
    return false;
  }
  if (filter.tags && !hasTags(tags, filter.tags)) {
    return false;
  }
  if (filter.since !== undefined && createdAt < filter.since) {
    return false;
  }
  if (filter.until !== undefined && createdAt > filter.until) {
    return false;
  }
  return true;
}

type Tag = { key: string; values: string[] };

function normalizeKinds(raw: unknown): Set<number> | undefined {
  if (raw === undefined || raw === null) {
    return undefined;
  }
  if (!Array.isArray(raw)) {
    throw new Error("kinds must be array");
  }
  return new Set(raw.map((value) => parseIntField(value, "kind")));
}

function normalizePubkeyPrefixes(raw: unknown): Uint8Array[] | undefined {
  if (raw === undefined || raw === null) {
    return undefined;
  }
  if (!Array.isArray(raw)) {
    throw new Error("pubkey_prefixes must be array");
  }
  return raw.map((value) => parseHexOrBytes(value, "pubkey_prefix", 16));
}

function normalizeTags(raw: unknown): Set<string> | undefined {
  if (raw === undefined || raw === null) {
    return undefined;
  }
  const tags = new Set<string>();
  if (Array.isArray(raw)) {
    for (const entry of raw) {
      if (!Array.isArray(entry) || entry.length !== 2) {
        throw new Error("tag filter entries must be [key, value]");
      }
      tags.add(`${String(entry[0])}:${String(entry[1])}`);
    }
    return tags;
  }
  if (typeof raw === "object") {
    for (const [key, values] of Object.entries(raw as Record<string, unknown>)) {
      if (!Array.isArray(values)) {
        throw new Error("tag values must be array");
      }
      for (const value of values) {
        tags.add(`${key}:${String(value)}`);
      }
    }
    return tags;
  }
  throw new Error("tags must be array or object");
}

function parseHexOrBytes(value: unknown, field: string, size: number): Uint8Array {
  if (value instanceof Uint8Array) {
    if (value.length !== size) {
      throw new Error(`${field} must be ${size} bytes`);
    }
    return value;
  }
  if (typeof value === "string") {
    const bytes = Uint8Array.from(Buffer.from(value, "hex"));
    if (bytes.length !== size) {
      throw new Error(`${field} must be ${size} bytes`);
    }
    return bytes;
  }
  throw new Error(`${field} must be hex string or bytes`);
}

function parseIntField(value: unknown, field: string): number {
  if (typeof value === "number") {
    return value;
  }
  if (typeof value === "string") {
    const parsed = Number(value);
    if (Number.isNaN(parsed)) {
      throw new Error(`${field} must be int`);
    }
    return parsed;
  }
  throw new Error(`${field} must be int`);
}

function parseTags(value: unknown): Tag[] {
  if (value === undefined || value === null) {
    return [];
  }
  if (!Array.isArray(value)) {
    throw new Error("tags must be array");
  }
  return value.map((entry) => {
    if (!Array.isArray(entry) || entry.length === 0) {
      throw new Error("invalid tag entry");
    }
    const [key, ...values] = entry;
    return { key: String(key), values: values.map((v) => String(v)) };
  });
}

function hasTags(tags: Tag[], required: Set<string>): boolean {
  const available = new Set<string>();
  for (const tag of tags) {
    for (const value of tag.values) {
      available.add(`${tag.key}:${value}`);
    }
  }
  for (const entry of required) {
    if (!available.has(entry)) {
      return false;
    }
  }
  return true;
}

function startsWith(value: Uint8Array, prefix: Uint8Array): boolean {
  if (prefix.length > value.length) {
    return false;
  }
  for (let i = 0; i < prefix.length; i += 1) {
    if (value[i] !== prefix[i]) {
      return false;
    }
  }
  return true;
}
